Introduce `UserStatisticsProvider` component and add support for respecting selected ruleset

frenzibyte opened the pull request:
- Closes https://github.com/ppy/osu/issues/26802
- [x] Depends on full-stack testing

I was initially leaning towards rewriting `SoloStatisticsWatcher` to become a general-purpose component surrounding around providing user statistics, but after [brief discussion](https://discord.com/channels/188630481301012481/188630652340404224/1205872037265014824) yesterday, I've settled on making a separate component for that instead, to keep complexity of each component to a minimum.

frenzibyte committed:
Introduce `LocalUserStatisticsProvider` component

frenzibyte committed:
Remove `IAPIProvider.Statistics` in favour of the new component

frenzibyte committed:
Update `UserRankPanel` implementation to use new component

frenzibyte committed:
Fix `ImportTest.TestOsuGameBase` having null ruleset

frenzibyte committed:
Fix `TestSceneUserPanel` tests failing

frenzibyte commented on osu.Game/Online/LocalUserStatisticsProvider.cs:
```diff
@@ -0,0 +1,94 @@
+                requestStatistics(ruleset.Value);
+        });
+
+        private void requestStatistics(RulesetInfo ruleset)
+        {
+            currentRequest = new GetUserRequest(api.LocalUser.Value.OnlineID, ruleset);
+            currentRequest.Success += u => statistics.Value = allStatistics[ruleset.ShortName] = u.Statistics;
+            api.Queue(currentRequest);
+        }
```
I did not emphasise too much about this, but I have mentioned in OP the following:
> This could be improved by fetching statistics of all rulesets once the user is logged in, but it would require performing four consecutive GetUserRequests for each ruleset to provide the statistics in full detail (i.e. when performing a GetUsersRequest, the RulesetsStatistics dictionary does not provide information about the country ranking in each ruleset).

`SoloStatisticsWatcher` is initialising user statistics for all rulesets using a `GetUsersRequest`, the response of that request does not include country ranking and potentially other attributes.

If anything, I would foresee this new component exposing statistics for all rulesets by a method or a dictionary, and `DifficultyRecommender` turning into an extension method, and `SoloStatisticsWatcher` becoming completely dependent on `LocalUserStatisticsProvider` and do away with the API request logic in the watcher. 

However, right now it's sort-of not pretty to achieve that, as I have to make the component perform one-time four consecutive API requests to receive full user statistics on each ruleset, and in complete honestly, I have no idea whether this is something that you would be fine with or not.

bdach replied:
at this point i'd probably ask if it is possible for us to receive full statistics in a single request because this feels wasteful.

@ppy/team-web do you think we could change some user-fetching endpoint to return complete rank data, including country ranking?

nanaya replied:
country ranking is rather expensive

frenzibyte replied:
> both `SoloStatisticsWatcher` and `DifficultyRecommender` do initial population of statistics for their own uses for _all rulesets_. so why is this doing the same in a _third_ place?
> 
> this is why i was floating potentially exposing the internal dictionary from `SoloStatisticsWatcher`.

Looking back at this, since the purpose of `LocalUserStatisticsProvider` is to track latest statistics and provide the latest one, I've made `UserStatisticsWatcher` fully rely on it and remove its local part of fetching initial statistics and tracking statistics in general.

frenzibyte commented on osu.Game/Online/LocalUserStatisticsProvider.cs:
```diff
@@ -0,0 +1,94 @@
+
+        private readonly Bindable<UserStatistics?> statistics = new Bindable<UserStatistics?>();
+
+        [Resolved]
+        private IBindable<RulesetInfo> ruleset { get; set; } = null!;
```
https://github.com/ppy/osu/pull/27128#discussion_r1486498229 is relevant.

peppy replied:
I don't really think that comment is relevant. The question here is more of an architectural one no? Like you can still cache the per-ruleset statistics, but instead of exposing the ruleset-specific bindable it would be up to the consumers (`DiscordRichPresence` and `UserRankPanel`).

*That said*, I think I'm okay with it existing as it is. At least until a use case comes up where we want to retrieve this for something that isn't the current global ruleset. Which arguable exists – the single usage you've piped through `GetStatisticsFor`. Is there a reason this flow doesn't do the online retrieval?

peppy replied:
I will say it does feel weird having this exist alongside `UserStatisticsWatcher` with data being awkwardly ferried between the two. Like they are both basically doing the same thing and the cross-communication is only required to keep the new component in sync?

frenzibyte replied:
For the sake of making each component have not but a single purpose, LocalUserStatisticsProvider and UserStatisticsWatcher have been split. The first exposes local user statistics to the game codebase, while the second has a very specific purpose tied to score submission which is to listen to submission events and display a change in user statistics while updating the former component with the new statistics.

I can understand that there's still a level of awkwardness between the two components since they both look up statistics online, and I wouldn't mind merging them together, but I would have to hear @bdach's opinion on that since he doesn't like that to be the case.

peppy replied:
I'm more focused on the fact that you added cross-talk between them, for the sake of saving one request I guess? If we're keeping both, I'd see that being removed.

frenzibyte replied:
The cross talk must exist for score submission to update the statistics exposed by `LocalUserStatisticsProvider`, the only way to avoid that is merging both components together.

bdach replied:
My position on this is unchanged: I would like the bindable to be gone.

The reasoning is extremely simple: when faced between two choices, one of which introduces ambient global state, and one of which does not, *always* choose the one that does not introduce the ambient global state unless you absolutely *have to*. It is simpler.

As I see it, `LocalUserStatisticsProvider` should be a dumb lookup facilitator / a _cache_. Same way something like `DifficultyCache` is. It should expose a dictionary for anyone to use in any way they see fit. Keyed by `(userId, rulesetId)`.

As for the interaction with `UserStatisticsWatcher` - `LocalUserStatisticsProvider` should expose a method to _invalidate and refetch_ data for a given cache key, and `UserStatistiscs` should use _that_ to retrieve its update.

frenzibyte replied:
Sounds good, will apply as proposed.

frenzibyte commented on osu.Game/OsuGameBase.cs:
```diff
@@ -328,7 +329,9 @@ private void load(ReadableKeyCombinationProvider keyCombinationProvider, Framewo
             dependencies.CacheAs(metadataClient = new OnlineMetadataClient(endpoints));
-            dependencies.CacheAs(soloStatisticsWatcher = new SoloStatisticsWatcher());
+
+            dependencies.CacheAs(localUserStatisticsProvider = new LocalUserStatisticsProvider());
+            dependencies.CacheAs(soloStatisticsWatcher = new SoloStatisticsWatcher(localUserStatisticsProvider));
```
Is this requesting something?

Joehuu commented:
Whenever this gets updated again, it should also be applied to discord rich presence as the rank is only fetched at game start / login and never updated.

frenzibyte committed:
Merge branch 'master' into user-statistics-provider

frenzibyte committed:
Fix post-merge errors

frenzibyte committed:
Move `LocalUserStatisticsProvider` to non-base game class and make dependency optional

frenzibyte committed:
Make `UserStatisticsWatcher` fully rely on `LocalUserStatisticsProvider`

frenzibyte committed:
Merge branch 'master' into user-statistics-provider

frenzibyte committed:
Update `DiscordRichPresence` to use new statistics provider component

frenzibyte commented:
> Whenever this gets updated again, it should also be applied to discord rich presence as the rank is only fetched at game start / login and never updated.

Should be addressed in https://github.com/ppy/osu/pull/27128/commits/663b769c710644cd12479b4a690cc24503f538bc

frenzibyte committed:
Reorder code slightly

frenzibyte commented:
Tested changes against a local full stack, works as expected.

peppy committed:
Merge branch 'master' into user-statistics-provider

frenzibyte committed:
Merge branch 'master' into user-statistics-provider

frenzibyte committed:
Decouple game-wide ruleset bindable and refactor `LocalUserStatisticsProvider`

This also throws away the logic of updating
`API.LocalUser.Value.Statistics`. Components should rely on
`LocalUserStatisticsProvider` instead for proper behaviour and ability
to update on statistics updates.

frenzibyte committed:
Make `DifficultyRecommender` rely on the statistics provider

frenzibyte committed:
Fix `UserRankPanel` not updating on ruleset change

frenzibyte commented:
As per the proposal in https://github.com/ppy/osu/pull/27128#discussion_r1842056283, the branch has taken another round, making `LocalUserStatisticsProvider` completely independent of the game-wide ruleset bindable's state, and acting purely as a cache for user statistics per ruleset.

frenzibyte commented on osu.Game/Online/LocalUserStatisticsProvider.cs:
```diff
@@ -0,0 +1,81 @@
+        /// A bindable communicating updates to the local user's statistics on any ruleset.
+        /// This does not guarantee the presence of old statistics, as it is invoked on initial population of statistics.
+        /// </summary>
+        public IBindable<UserStatisticsUpdate> StatisticsUpdate => statisticsUpdate;
```
This may seem odd at first but it's a comfortable method of communicating user statistics updates to any consuming component without going through the boring method of binding to an event, creating a callback function, and unbinding on disposal.

It's also shown to be directly useful for `DifficultyRecommender` to update its state with any update to any ruleset.

peppy replied:
I'm pretty sure @bdach's whole point was to not have this, but you've still got it present. Am I missing something.

frenzibyte replied:
No, it is about the component not tracking changes to the game-wide ruleset bindable. This bindable's main and sole purpose is to let any component be aware if a statistics has changed as a result of user logging out and logging in or submitting a new score. It is more of an event than a bindable but it's written as a bindable to make it easier to use rather than potentially fall in memory leak caused by not unsubscribing from the event.

peppy replied:
just make it an event if it's not accurate content wise....

frenzibyte replied:
Sure.

frenzibyte replied:
https://github.com/ppy/osu/pull/27128/commits/74daf85e489bb5402504729303b0f2dd1fa14e2f

frenzibyte commented on osu.Game/Online/ScoreBasedUserStatisticsUpdate.cs:
```diff
@@ -9,7 +9,7 @@ namespace osu.Game.Online
     /// Contains data about the change in a user's profile statistics after completing a score.
     /// </summary>
-    public class UserStatisticsUpdate
+    public class ScoreBasedUserStatisticsUpdate
```
This is a bit on the iffy side and it may not sit well with everyone, but with `LocalUserStatisticsProvider.StatisticsUpdate` requiring a model to contain change to user statistics on a specific ruleset, this class needed to be renamed to something else. The other option would be to pick a different name for the `StatisticsUpdate` model, but I think renaming this model is better since it is focused on scores anyway.

frenzibyte committed:
Only update user rank panel display when ruleset matches

Nothing behaviourally different, just reduce number of redundant calls.

frenzibyte committed:
Fix various test failures

frenzibyte committed:
Fix more test / component breakage

frenzibyte committed:
Replace bindable with an event

frenzibyte committed:
Handle logged out user

bdach commented on osu.Game/Online/UserStatisticsWatcher.cs:
```diff
@@ -98,31 +72,20 @@ private void userScoreProcessed(int userId, long scoreId)
 
-            latestStatistics.TryGetValue(rulesetName, out UserStatistics? latestRulesetStatistics);
-            latestRulesetStatistics ??= new UserStatistics();
+            if (update.OldStatistics != null)
+                latestUpdate.Value = new ScoreBasedUserStatisticsUpdate(scorePendingUpdate, update.OldStatistics, update.NewStatistics);
```
I find this quite dodgy. What's the guarantee that you get the correct statistics update here and not some unrelated update queued by someone else later? Or that some unrelated statistics update isn't *treated* as a "score based" stats update?

If you ask me, `LocalUserStatisticsProvider.RefetchStatistics()` should return a `Task<UserStatisticsUpdate>`, so that task can be stored, and this component can be _sure_ that it's handling the data it explicitly requested.

frenzibyte replied:
Since we want to be aware of which update is tied to the requested statistics refetch, wouldn't it be more easy to let the function accept a callback and invoke it directly from within the `Success` event of the request? I understand this heavily involves delegate variable capturing, but I think it's still within reason since requests are transient and die immediately once they're fulfilled, so there shouldn't be any room for potential event leaks? I'm not sure, I might be severely misunderstanding this. Here's an example diff of my proposal:
```diff
diff --git a/osu.Game/Online/LocalUserStatisticsProvider.cs b/osu.Game/Online/LocalUserStatisticsProvider.cs
index a25f5b05aa..1d322614e1 100644
--- a/osu.Game/Online/LocalUserStatisticsProvider.cs
+++ b/osu.Game/Online/LocalUserStatisticsProvider.cs
@@ -3,7 +3,9 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
 using System.Linq;
+using System.Threading.Tasks;
 using osu.Framework.Allocation;
 using osu.Framework.Graphics;
 using osu.Game.Extensions;
@@ -36,7 +38,6 @@ public partial class LocalUserStatisticsProvider : Component
         private IAPIProvider api { get; set; } = null!;
 
         private readonly Dictionary<string, UserStatistics> statisticsCache = new Dictionary<string, UserStatistics>();
-        private readonly Dictionary<string, GetUserRequest> statisticsRequests = new Dictionary<string, GetUserRequest>();
 
         /// <summary>
         /// Returns the <see cref="UserStatistics"/> currently available for the given ruleset.
@@ -62,23 +63,21 @@ private void initialiseStatistics()
                 RefetchStatistics(ruleset);
         }
 
-        public void RefetchStatistics(RulesetInfo ruleset)
+        public void RefetchStatistics(RulesetInfo ruleset, Action<UserStatisticsUpdate>? callback = null)
         {
-            if (statisticsRequests.TryGetValue(ruleset.ShortName, out var previousRequest))
-                previousRequest.Cancel();
-
-            var request = statisticsRequests[ruleset.ShortName] = new GetUserRequest(api.LocalUser.Value.Id, ruleset);
-            request.Success += u => UpdateStatistics(u.Statistics, ruleset);
+            var request = new GetUserRequest(api.LocalUser.Value.Id, ruleset);
+            request.Success += u => UpdateStatistics(u.Statistics, ruleset, callback);
             api.Queue(request);
         }
 
-        protected void UpdateStatistics(UserStatistics newStatistics, RulesetInfo ruleset)
+        protected void UpdateStatistics(UserStatistics newStatistics, RulesetInfo ruleset, Action<UserStatisticsUpdate>? callback = null)
         {
             var oldStatistics = statisticsCache.GetValueOrDefault(ruleset.ShortName);
-
-            statisticsRequests.Remove(ruleset.ShortName);
             statisticsCache[ruleset.ShortName] = newStatistics;
-            StatisticsUpdated?.Invoke(new UserStatisticsUpdate(ruleset, oldStatistics, newStatistics));
+
+            var update = new UserStatisticsUpdate(ruleset, oldStatistics, newStatistics);
+            callback?.Invoke(update);
+            StatisticsUpdated?.Invoke(update);
         }
     }
 
diff --git a/osu.Game/Online/UserStatisticsWatcher.cs b/osu.Game/Online/UserStatisticsWatcher.cs
index 8ed1ff594d..e9b7c8e652 100644
--- a/osu.Game/Online/UserStatisticsWatcher.cs
+++ b/osu.Game/Online/UserStatisticsWatcher.cs
@@ -2,8 +2,11 @@
 // See the LICENCE file in the repository root for full licence text.
 
 using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading.Tasks;
 using osu.Framework.Allocation;
 using osu.Framework.Bindables;
+using osu.Framework.Extensions;
 using osu.Framework.Extensions.ObjectExtensions;
 using osu.Framework.Graphics;
 using osu.Game.Extensions;
@@ -23,8 +26,6 @@ public partial class UserStatisticsWatcher : Component
         public IBindable<ScoreBasedUserStatisticsUpdate?> LatestUpdate => latestUpdate;
         private readonly Bindable<ScoreBasedUserStatisticsUpdate?> latestUpdate = new Bindable<ScoreBasedUserStatisticsUpdate?>();
 
-        private ScoreInfo? scorePendingUpdate;
-
         [Resolved]
         private SpectatorClient spectatorClient { get; set; } = null!;
 
@@ -43,7 +44,6 @@ protected override void LoadComplete()
             base.LoadComplete();
 
             spectatorClient.OnUserScoreProcessed += userScoreProcessed;
-            statisticsProvider.StatisticsUpdated += onStatisticsUpdated;
         }
 
         /// <summary>
@@ -72,21 +72,13 @@ private void userScoreProcessed(int userId, long scoreId)
             if (!watchedScores.Remove(scoreId, out var scoreInfo))
                 return;
 
-            scorePendingUpdate = scoreInfo;
-            statisticsProvider.RefetchStatistics(scoreInfo.Ruleset);
+            statisticsProvider.RefetchStatistics(scoreInfo.Ruleset, u => Schedule(() =>
+            {
+                if (u.OldStatistics != null)
+                    latestUpdate.Value = new ScoreBasedUserStatisticsUpdate(scoreInfo, u.OldStatistics, u.NewStatistics);
+            }));
         }
 
-        private void onStatisticsUpdated(UserStatisticsUpdate update) => Schedule(() =>
-        {
-            if (scorePendingUpdate == null || !update.Ruleset.Equals(scorePendingUpdate.Ruleset))
-                return;
-
-            if (update.OldStatistics != null)
-                latestUpdate.Value = new ScoreBasedUserStatisticsUpdate(scorePendingUpdate, update.OldStatistics, update.NewStatistics);
-
-            scorePendingUpdate = null;
-        });
-
         protected override void Dispose(bool isDisposing)
         {
             if (spectatorClient.IsNotNull())
```

I didn't go for it at first because I didn't think I need to care about which update is which in `UserStatisticsWatcher`, but here we are.

bdach replied:
I suppose a callback is the same thing. I'm not sure I have a strong preference of one over the other.

bdach commented on osu.Game/Online/API/Requests/Responses/APIUser.cs:
```diff
@@ -223,8 +223,10 @@ public class KudosuCount
         /// User statistics for the requested ruleset (in the case of a <see cref="GetUserRequest"/> or <see cref="GetFriendsRequest"/> response).
-        /// Otherwise empty.
         /// </summary>
+        /// <remarks>
+        /// This returns null when accessed from <see cref="IAPIProvider.LocalUser"/>. Use <see cref="LocalUserStatisticsProvider"/> instead.
+        /// </remarks>
```
oh how i *love* this model class (heavy sarcasm)

bdach commented on osu.Game/Online/API/IAPIProvider.cs:
```diff
@@ -29,11 +29,6 @@ public interface IAPIProvider
-        /// <summary>
-        /// The current user's online statistics.
-        /// </summary>
-        IBindable<UserStatistics?> Statistics { get; }
```
For one, very happy to see this gone from here

frenzibyte committed:
Replace event subscription with callback in `UserStatisticsWatcher`

Also no longer cancels previous API requests as there's no actual need to do it.

frenzibyte committed:
Enable NRT and fix code

frenzibyte committed:
Fix failing test

bdach committed:
Add guard against fetching statistics for non-legacy rulesets

bdach reviewed (dismissed):
added one precautionary guard (https://github.com/ppy/osu/pull/27128/commits/0a3f3c3210dd50f5a50c7fb0df875299abc9ffe8) but otherwise seems fine

peppy commented:
Will take one final look at this.

peppy commented:
Statistics don't load for me from a fresh startup:

![osu! 2024-11-26 at 03 30 31](https://github.com/user-attachments/assets/044d13e9-6c80-4fa6-8936-871c81274854)

peppy commented:
Here's the failing flow:

[runtime] 2024-11-26 03:48:47 [verbose]: STATS: update from 1 to 2
[runtime] 2024-11-26 03:48:47 [verbose]: STATS: init
[runtime] 2024-11-26 03:48:47 [verbose]: STATS: refreshed
[network] 2024-11-26 03:48:47 [verbose]: Failing request osu.Game.Online.API.Requests.GetUserRequest (System.Net.WebException: User not logged in)
[network] 2024-11-26 03:48:47 [verbose]: Failing request osu.Game.Online.API.Requests.GetUserRequest (System.Net.WebException: User not logged in)
[network] 2024-11-26 03:48:47 [verbose]: Failing request osu.Game.Online.API.Requests.GetUserRequest (System.Net.WebException: User not logged in)
[network] 2024-11-26 03:48:47 [verbose]: Failing request osu.Game.Online.API.Requests.GetUserRequest (System.Net.WebException: User not logged in)

```diff
diff --git a/osu.Game/Online/API/APIAccess.cs b/osu.Game/Online/API/APIAccess.cs
index ec48fa2436..45daaaf046 100644
--- a/osu.Game/Online/API/APIAccess.cs
+++ b/osu.Game/Online/API/APIAccess.cs
@@ -339,12 +339,11 @@ private void attemptConnect()
 
                     userReq.Success += me =>
                     {
-                        me.Status.Value = configStatus.Value ?? UserStatus.Online;
-
-                        setLocalUser(me);
-
                         state.Value = me.SessionVerified ? APIState.Online : APIState.RequiresSecondFactorAuth;
                         failureCount = 0;
+
+                        me.Status.Value = configStatus.Value ?? UserStatus.Online;
+                        setLocalUser(me);
                     };
 
                     if (!handleRequest(userReq))

```

This may fix but also may break others things. Chicken-egg scenario with state and localuser bindables.

peppy reviewed (changes requested):
🙈

frenzibyte commented:
Sounds like the perfect solution would be to schedule. If the status is updated directly after the local user is updated, then all that should be done is to just schedule the callback so enough time is passed for other states to be updated.

peppy commented:
Sounds like the Perfect Isolated Hacky Fix.

frenzibyte commented:
Oh yeah sure let's reinvent the whole `IAPIProvider` system...; I'll push a schedule fix. It's already done in `UserStatisticsWatcher` matter of fact.

frenzibyte committed:
Schedule the thing

Queuing up requests on change to `api.LocalUser` is bad because the API
state is updated after `LocalUser` is updated, therefore we have to
schhhhhedullllllllleeeeeeeeeeeeeeee.

peppy commented:
If you're adding a schedule you better have inline documentation explaining why, so the next person who actually wants to fix the issue knows the context.

frenzibyte committed:
Add inline comment

peppy commented:
@bdach please check you're okay with the schedule usage.

bdach commented:
I can live with the schedule. It's a hacky fix but at least it's isolated and it'd be nice to actually get this PR off the list one way or another after 9 months.

> It's already done in `UserStatisticsWatcher` matter of fact.

Baffled by this remark though, because _the reason why that component schedules is completely different_. There it's for _basic thread safety_. The primary usage of the watcher flow is in online score submission, which is _async_. That's not working around bindable idiosyncrasies like this is.

frenzibyte commented:
> > It's already done in `UserStatisticsWatcher` matter of fact.
> 
> Baffled by this remark though, because _the reason why that component schedules is completely different_. There it's for _basic thread safety_. The primary usage of the watcher flow is in online score submission, which is _async_. That's not working around bindable idiosyncrasies like this is.

No? Check `master`, I'm referring to the schedule in `onUserChanged`, that method is used nowhere other than the `api.LocalUser` bindable callback.

bdach commented:
> I'm referring to the schedule in `onUserChanged`,

My bad I guess. I don't even remember why that one is there, and I wrote that code. I would have liked to think I would have mentioned if the reason for the schedule was the same, or I guess past me just sucks 🤷
